File: src/main/java/io/Cbuild/gnu_make_functions/compiler/compile_fn.java
package io.Cbuild.gnu_make_functions.compiler;

import io.Cbuild.cBuildIR;
import io.Cbuild.cbuildBaseVisitor;
import io.Cbuild.cbuildException;
import io.Cbuild.cbuildParser;
import io.Cbuild.gnu_make_functions.make_function_dispatcher;
import io.Cbuild.gnu_make_functions.type;
import org.antlr.v4.runtime.tree.ParseTree;


public class compile_fn extends cbuildBaseVisitor<cBuildIR.ValueIR > implements I_compiler_base {


    private final make_function_dispatcher dispatcher;

    public compile_fn() {
        this.dispatcher = new make_function_dispatcher();
    }

    @Override
    public cBuildIR.FunctionIR compile(cbuildParser.FunctionContext ctx, type.MakeFunction function) {
        int given =  ctx.arguments() == null ? 0 : ctx.arguments().argument().size();
        int expected = function.arity();

        if (given != expected) {
            throw new cbuildException(
                    cbuildException.ErrorType.SEMANTIC,
                    "function '" + function.getFnName() + "' expects " +
                            expected + " argument(s), but got " + given,
                    ctx.start.getLine(),
                    ctx.start.getCharPositionInLine() + 1
            );
        }

        cBuildIR.FunctionIR fn = new cBuildIR.FunctionIR(function.getFnName());

        for (cbuildParser.ArgumentContext argCtx : ctx.arguments().argument()) {
            fn.args.add(visitArgument(argCtx));
        }

        return fn;
    }

    @Override
    public cBuildIR.ValueIR visitArgument(cbuildParser.ArgumentContext ctx) {
        return visitExpressions(ctx.expressions());
    }

    @Override
    public cBuildIR.ValueIR visitExpressions(cbuildParser.ExpressionsContext ctx) {
        cBuildIR.ValueIR value = new cBuildIR.ValueIR();

        for (ParseTree ast_node : ctx.children) {
            cBuildIR.ValueIR partValue = ast_node.accept(this);
            value.parts.addAll(partValue.parts);
        }

        return value;
    }

    @Override
    public cBuildIR.ValueIR visitExpression(cbuildParser.ExpressionContext ctx) {
        cBuildIR.ValueIR value = new cBuildIR.ValueIR();

        for (cbuildParser.Expression_atomContext atomCtx : ctx.expression_atom()) {
            cBuildIR.ValueIR atomValue = visitExpression_atom(atomCtx);
            value.parts.addAll(atomValue.parts);
        }

        return value;
    }

    @Override
    public cBuildIR.ValueIR visitExpression_atom(cbuildParser.Expression_atomContext ctx) {
        cBuildIR.ValueIR value = new cBuildIR.ValueIR();

        if (ctx.text() != null) {
            value.parts.add(new cBuildIR.TextPart(ctx.text().getText()));
            return value;
        }

        if (ctx.exprs_nested() != null) {
            cBuildIR.ValueIR nameExpr = visitExprs_nested(ctx.exprs_nested());
            value.parts.add(new cBuildIR.VarRefPart(nameExpr));
            return value;
        }

        if (ctx.function() != null) {
            if(this.dispatcher.has(ctx.function().function_name().getText())) {
                make_function_dispatcher.MakeFunctionHandler handler =
                        this.dispatcher.getHandler(ctx.function().function_name().getText());
                cBuildIR.FunctionIR functionIR = (cBuildIR.FunctionIR ) handler.compile(ctx.function());
                value.parts.add(new cBuildIR.FunctionCallPart(functionIR));
            }
            else {
                cBuildIR.ValueIR varName = new cBuildIR.ValueIR();
                varName.parts.add(new cBuildIR.TextPart(ctx.function().function_name().getText()));
                value.parts.add(new cBuildIR.VarRefPart(varName));
            }

            return value;
        }

        return value;
    }

    @Override
    public cBuildIR.ValueIR visitExprs_nested(cbuildParser.Exprs_nestedContext ctx) {
        cBuildIR.ValueIR value = new cBuildIR.ValueIR();

        for (cbuildParser.Expr_nestedContext nestedCtx : ctx.expr_nested()) {
            cBuildIR.ValueIR nestedValue = visitExpr_nested(nestedCtx);
            value.parts.addAll(nestedValue.parts);
        }

        return value;
    }

    @Override
    public cBuildIR.ValueIR visitExpr_nested(cbuildParser.Expr_nestedContext ctx) {
        cBuildIR.ValueIR value = new cBuildIR.ValueIR();

        for (cbuildParser.Expr_nested_atomContext atomCtx : ctx.expr_nested_atom()) {
            cBuildIR.ValueIR atomValue = visitExpr_nested_atom(atomCtx);
            value.parts.addAll(atomValue.parts);
        }

        return value;
    }

    @Override
    public cBuildIR.ValueIR visitExpr_nested_atom(cbuildParser.Expr_nested_atomContext ctx) {
        cBuildIR.ValueIR value = new cBuildIR.ValueIR();

        if (ctx.text_nested() != null) {
            value.parts.add(new cBuildIR.TextPart(ctx.text_nested().getText()));
            return value;
        }

        if (ctx.exprs_nested() != null) {
            cBuildIR.ValueIR nameExpr = visitExprs_nested(ctx.exprs_nested());
            value.parts.add(new cBuildIR.VarRefPart(nameExpr));
            return value;
        }

        if (ctx.function() != null) {

            if(this.dispatcher.has(ctx.function().function_name().getText())) {
                make_function_dispatcher.MakeFunctionHandler handler =
                        this.dispatcher.getHandler(ctx.function().function_name().getText());
                cBuildIR.FunctionIR functionIR = (cBuildIR.FunctionIR ) handler.compile(ctx.function());
                value.parts.add(new cBuildIR.FunctionCallPart(functionIR));
            }
            else {
                cBuildIR.ValueIR varName = new cBuildIR.ValueIR();
                varName.parts.add(new cBuildIR.TextPart(ctx.function().function_name().getText()));
                value.parts.add(new cBuildIR.VarRefPart(varName));
            }

            return value;
        }

        return value;
    }

    @Override
    public cBuildIR.ValueIR visitWs(cbuildParser.WsContext ctx) {
        cBuildIR.TextPart part = new cBuildIR.TextPart(ctx.getText());
        cBuildIR.ValueIR val = new cBuildIR.ValueIR();
        val.parts.add(part);
        return val;
    }

}

File: src/gnu-make-functions/compiler/compile_fn.ts
import { AbstractParseTreeVisitor } from "antlr4ts/tree/AbstractParseTreeVisitor";
import { cbuildVisitor } from "@parser/cbuildVisitor";
import {
  ArgumentContext,
  Expr_nested_atomContext,
  Expr_nestedContext,
  Expression_atomContext,
  ExpressionContext,
  ExpressionsContext,
  Exprs_nestedContext,
  FunctionContext,
  WsContext,
} from "@parser/cbuildParser";
import {
  FunctionIR,
  ValueIR,
  functionCallPart,
  textPart,
  varRefPart,
} from "@compiler/ir";
import { cbuildException, ErrorType } from "@src/cbuild-exception";
import { make_function_dispatcher } from "@gnu-make-functions/make_function_dispatcher";
import type { MakeFunction } from "@gnu-make-functions/type";
import { I_compiler_base } from "@src/gnu-make-functions/compiler/Icompiler-base";

export class compile_fn
  extends AbstractParseTreeVisitor<ValueIR>
  implements I_compiler_base, cbuildVisitor<ValueIR>
{
  private readonly dispatcher: make_function_dispatcher;

  public constructor() {
    super();
    this.dispatcher = new make_function_dispatcher();
  }

  protected defaultResult(): ValueIR {
    return new ValueIR();
  }

  public compile(ctx: FunctionContext, func: MakeFunction): FunctionIR {
    const given =
      ctx.arguments() == null ? 0 : ctx.arguments()!.argument().length;
    const expected = func.arity();

    if (given !== expected) {
      throw new cbuildException(
        ErrorType.SEMANTIC,
        "function '" +
          func.getFnName() +
          "' expects " +
          expected +
          " argument(s), but got " +
          given,
        ctx.start.line,
        ctx.start.charPositionInLine + 1,
      );
    }

    const fn = new FunctionIR(func.getFnName());

    for (const argCtx of ctx.arguments()!.argument()) {
      fn.args.push(this.visitArgument(argCtx));
    }

    return fn;
  }

  public visitArgument(ctx: ArgumentContext): ValueIR {
    return this.visitExpressions(ctx.expressions());
  }

  public visitExpressions(ctx: ExpressionsContext): ValueIR {
    const value = new ValueIR();

    for (const ast_node of ctx.children ?? []) {
      const partValue = ast_node.accept(this);
      value.parts.push(...partValue.parts);
    }

    return value;
  }

  public visitExpression(ctx: ExpressionContext): ValueIR {
    const value = new ValueIR();

    for (const atomCtx of ctx.expression_atom()) {
      const atomValue = this.visitExpression_atom(atomCtx);
      value.parts.push(...atomValue.parts);
    }

    return value;
  }

  public visitExpression_atom(ctx: Expression_atomContext): ValueIR {
    const value = new ValueIR();

    if (ctx.text() != null) {
      value.parts.push(textPart(ctx.text()!.text));
      return value;
    }

    if (ctx.exprs_nested() != null) {
      const nameExpr = this.visitExprs_nested(ctx.exprs_nested()!);
      value.parts.push(varRefPart(nameExpr));
      return value;
    }

    if (ctx.function() != null) {
      if (this.dispatcher.has(ctx.function()!.function_name()!.text)) {
        const handler = this.dispatcher.getHandler(
          ctx.function()!.function_name()!.text,
        );
        const functionIR = handler.compile(ctx.function()!);
        value.parts.push(functionCallPart(functionIR));
      } else {
        const varName = new ValueIR();
        varName.parts.push(textPart(ctx.function()!.function_name()!.text));
        value.parts.push(varRefPart(varName));
      }

      return value;
    }

    return value;
  }

  public visitExprs_nested(ctx: Exprs_nestedContext): ValueIR {
    const value = new ValueIR();

    for (const nestedCtx of ctx.expr_nested()) {
      const nestedValue = this.visitExpr_nested(nestedCtx);
      value.parts.push(...nestedValue.parts);
    }

    return value;
  }

  public visitExpr_nested(ctx: Expr_nestedContext): ValueIR {
    const value = new ValueIR();

    for (const atomCtx of ctx.expr_nested_atom()) {
      const atomValue = this.visitExpr_nested_atom(atomCtx);
      value.parts.push(...atomValue.parts);
    }

    return value;
  }

  public visitExpr_nested_atom(ctx: Expr_nested_atomContext): ValueIR {
    const value = new ValueIR();

    if (ctx.text_nested() != null) {
      value.parts.push(textPart(ctx.text_nested()!.text));
      return value;
    }

    if (ctx.exprs_nested() != null) {
      const nameExpr = this.visitExprs_nested(ctx.exprs_nested()!);
      value.parts.push(varRefPart(nameExpr));
      return value;
    }

    if (ctx.function() != null) {
      if (this.dispatcher.has(ctx.function()!.function_name()!.text)) {
        const handler = this.dispatcher.getHandler(
          ctx.function()!.function_name()!.text,
        );
        const functionIR = handler.compile(ctx.function()!);
        value.parts.push(functionCallPart(functionIR));
      } else {
        const varName = new ValueIR();
        varName.parts.push(textPart(ctx.function()!.function_name()!.text));
        value.parts.push(varRefPart(varName));
      }

      return value;
    }

    return value;
  }

  public visitWs(ctx: WsContext): ValueIR {
    const part = textPart(ctx.text);
    const val = new ValueIR();
    val.parts.push(part);
    return val;
  }
}
